File: apps/api/routers/library/folders.py
"""Library folders — manual corpus organisation (feature-corpus-folders.md, ADR-025 F1).

Flat named folders with bulk membership. Since F2 a folder can also scope one chat turn's
retrieval (``POST /api/chat`` ``scope_folder_id``), so these writes are visible on the answer
path — but only by id: membership is resolved per turn, never cached here.

Deleting a folder never deletes documents (spec D6).
"""

from __future__ import annotations

from fastapi import APIRouter, HTTPException

from apps.api.models.folders import (
    FolderCreate,
    FolderMembers,
    FolderRename,
    LibraryFolderPayload,
)

router = APIRouter()


@router.get("/api/library/folders")
def list_folders_route() -> list[LibraryFolderPayload]:
    """Every folder with its non-archived member count (feature-corpus-folders.md, ADR-025 F1).

    A folder organises the Library and, since F2, can scope one chat turn's retrieval
    (POST /api/chat `scope_folder_id`)."""
    from doc_assistant.library import list_folders

    return [LibraryFolderPayload.from_folder(f) for f in list_folders()]


@router.post("/api/library/folders")
def create_folder_route(body: FolderCreate) -> LibraryFolderPayload:
    """Create a folder. Idempotent on the case-folded name (returns the existing one)."""
    from doc_assistant.library import create_folder

    try:
        folder = create_folder(body.name, body.description)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e)) from e
    return LibraryFolderPayload.from_folder(folder)


@router.patch("/api/library/folders/{folder_id}")
def rename_folder_route(folder_id: str, body: FolderRename) -> LibraryFolderPayload:
    """Rename a folder. 404 if unknown, 400 if the name collides with another folder."""
    from doc_assistant.library import rename_folder

    try:
        folder = rename_folder(folder_id, body.name)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e)) from e
    if folder is None:
        raise HTTPException(status_code=404, detail="folder not found")
    return LibraryFolderPayload.from_folder(folder)


@router.post("/api/library/folders/{folder_id}/documents")
def add_folder_documents_route(folder_id: str, body: FolderMembers) -> LibraryFolderPayload:
    """Add documents to a folder (bulk, idempotent; unknown ids are skipped). 404 if
    the folder is unknown."""
    from doc_assistant.library import add_documents_to_folder

    folder = add_documents_to_folder(folder_id, body.document_ids)
    if folder is None:
        raise HTTPException(status_code=404, detail="folder not found")
    return LibraryFolderPayload.from_folder(folder)


@router.delete("/api/library/folders/{folder_id}/documents/{doc_id}")
def remove_folder_document_route(folder_id: str, doc_id: str) -> LibraryFolderPayload:
    """Remove one document from a folder (a no-op if it isn't a member). The document
    itself is untouched. 404 if the folder is unknown."""
    from doc_assistant.library import remove_documents_from_folder

    folder = remove_documents_from_folder(folder_id, [doc_id])
    if folder is None:
        raise HTTPException(status_code=404, detail="folder not found")
    return LibraryFolderPayload.from_folder(folder)


@router.delete("/api/library/folders/{folder_id}")
def delete_folder_route(folder_id: str) -> dict[str, bool]:
    """Delete a folder. Never deletes documents (spec D6). 404 if unknown."""
    from doc_assistant.library import delete_folder

    if not delete_folder(folder_id):
        raise HTTPException(status_code=404, detail="folder not found")
    return {"ok": True}

File: apps/desktop/src/lib/core/api/folders.ts
// Thin `folders` client for the desktop API — fetch + parsing only, no business logic.
// Pairs with apps/api/routers/folders.py and apps/api/models/folders.py; see
// docs/architecture.md, section "apps/ — the domain spine".

import { API_BASE, errorDetail } from './_base'
import type {
  LibraryFolder,
} from '../types'

export async function listFolders(): Promise<LibraryFolder[]> {
  const r = await fetch(`${API_BASE}/api/library/folders`)
  if (!r.ok) throw new Error(await errorDetail(r, 'folders'))
  return (await r.json()) as LibraryFolder[]
}
/** Create a folder. Idempotent on the case-folded name — the existing folder comes back. */
export async function createFolder(name: string): Promise<LibraryFolder> {
  const r = await fetch(`${API_BASE}/api/library/folders`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ name }),
  })
  if (!r.ok) throw new Error(await errorDetail(r, 'create folder'))
  return (await r.json()) as LibraryFolder
}
export async function renameFolder(folderId: string, name: string): Promise<LibraryFolder> {
  const r = await fetch(`${API_BASE}/api/library/folders/${encodeURIComponent(folderId)}`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ name }),
  })
  if (!r.ok) throw new Error(await errorDetail(r, 'rename folder'))
  return (await r.json()) as LibraryFolder
}
/** Delete the folder only — the documents in it are never touched (spec D6). */
export async function deleteFolder(folderId: string): Promise<void> {
  const r = await fetch(`${API_BASE}/api/library/folders/${encodeURIComponent(folderId)}`, {
    method: 'DELETE',
  })
  if (!r.ok) throw new Error(await errorDetail(r, 'delete folder'))
}
/** Bulk-add documents to a folder. Idempotent; unknown ids are skipped server-side. */
export async function addDocumentsToFolder(
  folderId: string,
  documentIds: string[],
): Promise<LibraryFolder> {
  const r = await fetch(
    `${API_BASE}/api/library/folders/${encodeURIComponent(folderId)}/documents`,
    {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ document_ids: documentIds }),
    },
  )
  if (!r.ok) throw new Error(await errorDetail(r, 'add documents to folder'))
  return (await r.json()) as LibraryFolder
}
export async function removeDocumentFromFolder(
  folderId: string,
  documentId: string,
): Promise<LibraryFolder> {
  const r = await fetch(
    `${API_BASE}/api/library/folders/${encodeURIComponent(folderId)}/documents/${encodeURIComponent(documentId)}`,
    { method: 'DELETE' },
  )
  if (!r.ok) throw new Error(await errorDetail(r, 'remove document from folder'))
  return (await r.json()) as LibraryFolder
}

// Tag families (feature-tag-families.md, PR-1). A family collapses near-duplicate keywords
// (`llm`/`llms`) into one filterable entry — errors carry the backend's `detail` (400/404).
